### PR TITLE
CryptoJS

### PR DESCRIPTION
**Pull request by jtormey**
Replaces deprecated CryptoJS with crypto-browserify (used by BitcoinJS).

@kristovatlas

### CODE REVIEW

**Comment by pernas** [comment]
ack, thanks @jtormey

**Comment by Sjors** [comment]
The [Crypto](https://nodejs.org/api/crypto.html#crypto_crypto) module is baked into NodeJS. [crypto-browserify](https://github.com/crypto-browserify) is a collection of browser optimized drop-in replacements for many of it's methods. If you use crypto directly our NodeJS wallet will use use the built-in module and our web wallet runs it through Browserify. It's important to note that Browserify does not produce the exact same code as crypto-browserify. E.g.:

```js
var crypto = require('crypto');
var cipher = crypto.createCipheriv(...)
```
Is not the the same code in the browser as:
```js
var crypto = require("crypto-browserify");
var cipher = crypto.createCipheriv(...);
```

To see this, compare `dist/wallet.js` in this PR to [createCipheriv in browserify-cipher](https://github.com/crypto-browserify/browserify-cipher/blob/master/browser.js#L37).

I don't know if the manually curated crypto-browserify functions are worth using instead of the automatically browserified ones, when there are no issues. Perhaps they perform better and have better cross-browser support. On the other hand perhaps they are not as well maintained and screened as the originals.

I do believe we should be consistent for each specific function. We're using [crypto-browserify/randombytes](https://www.npmjs.com/package/randombytes) in our [transaction builder](https://github.com/blockchain/My-Wallet-V3/blob/master/src/transaction.js#L5), but this PR uses `crypto.randomBytes` (which browserifies the built-in Node function). For that reason I [NACK](https://github.com/bitcoin/bitcoin/issues/6100) until then.

For consistently it might make sense to replace [crypto.pbkdf2Sync](https://github.com/crypto-browserify/pbkdf2), [crypto.createHash](https://github.com/crypto-browserify/createHash) (already whitelisted), [createCipheriv](https://github.com/crypto-browserify/browserify-cipher/) with the crypto-browserify variants.

One more thing, a small issue: `import-export.js` still uses `var hash256 = Bitcoin.crypto.hash256;`.

**Comment by Sjors** [comment]
I would also like to see @mpfluger's blessing on the rather large changes to  ImportExport.Crypto_scrypt,  encryptDataWithPassword, stretchPassword and decryptWalletV1. I manually tested in the browser one BIP 38 vector, changing password, turning 2nd password on and off and backups. I haven't tested opening a v1 wallet.

**Comment by jtormey** [comment]
@Sjors, thanks for looking this over. To address some of you concerns:

Requiring 'crypto' versus 'crypto-browserify' both return the same code in the browser. Browserify cannot actually "browserify" native code, so it replaces it with crypto-browserify (in fact, you can see that 'crypto-browserify' is a dependency of browserify). Requiring 'crypto' has the added benefit of using native functions in node

`crypto.randomBytes` and `cryptoBrowserify.randomBytes` are the same, for the same reason as above. In our transaction builder, we use a separate module ('randombytes') for some reason.

The BitcoinJS hash function is ok to use, since that library isn't the one that was removed. But if we want to change it it's not a big deal. `Bitcoin.crypto.hash256(x) == sha256(sha256(x))`

**Comment by Sjors** [comment]
Ah, now I see the magic performed by [built-ins](https://github.com/substack/browserify-handbook#builtins). So it's really not necessary to list `randombytes` in dependencies, since `crypto.randomBytes` will automatically be replaced by `browserify/randombytes`? @pernas?

In that case ACK.

**Comment by mpfluger** [comment]
ACK

We have tests for all the old wallet formats here:

https://github.com/blockchain/My-Wallet-V3/blob/a435713334fe64af0d8dce467f7f43654da5668e/tests/wallet_spec.js.coffee#L59

So if the tests pass, there shouldn't be any problem with legacy wallet format decryption.

I'm not 100% sure on the randombytes question from above. This is the crypto-browserify version, so Sjors may be right that we can simply leave it out of the `package.json` and it will keep working.

**Comment by Sjors** [comment]
Although these tests look solid, I would not blindly rely on them without a counter example such as "should fail to decrypt a corrupted legacy v1 wallet with ...".

If we remove it from package.json we also have to replace references to use `crypto.randomBytes`.

**Comment by Sjors** [comment]
Also a large chunk of tests are still disabled (this one isn't, I just checked).

**Comment by pernas** [comment]
i am not sure what to answer on the randombytes question either

**Comment by kristovatlas** [comment]
> I would not blindly rely on them without a counter example such as "should fail to decrypt a corrupted legacy v1 wallet with ...".

+1

File: check-dependencies.rb [hunk 1]
```diff
@@ -172,69 +172,40 @@ def check_commits!(deps, whitelist, output_deps, type)
 #########
 # NPM   #
 #########
-if package["name"] == "My-Wallet-HD" # Only My-Wallet-HD uses NPM
 
-  shrinkwrap = JSON.parse(File.read('npm-shrinkwrap.json'))
```
Comment (Mata2580) [inline check-dependencies.rb]: 26

**Pull request merged**